File: common/src/main/java/com/imsproject/common/dataAccess/abstracts/CounterDAOBase.java
package com.imsproject.common.dataAccess.abstracts;


import com.imsproject.common.dataAccess.CreateTableQueryBuilder;
import com.imsproject.common.dataAccess.DaoException;
import com.imsproject.common.dataAccess.OfflineResultSet;

import java.sql.SQLException;

public abstract class CounterDAOBase implements CounterDAO {

    private final SQLExecutor cursor;
    private final String TABLE_NAME;
    private final String COLUMN_NAME;

    protected CounterDAOBase(SQLExecutor cursor, String tableName, String columnName) throws DaoException {
        this.cursor = cursor;
        TABLE_NAME = tableName;
        COLUMN_NAME = columnName;
        initTable();
    }

    protected void initTable() throws DaoException {
        CreateTableQueryBuilder createTableQueryBuilder = CreateTableQueryBuilder.create(TABLE_NAME);

        createTableQueryBuilder.addColumn(COLUMN_NAME,
                CreateTableQueryBuilder.ColumnType.INTEGER,
                CreateTableQueryBuilder.ColumnModifier.NOT_NULL);

        String query = createTableQueryBuilder.build();
        try {
            cursor.executeWrite(query);
            query = String.format("SELECT * FROM %s;", TABLE_NAME);
            if(cursor.executeRead(query).isEmpty()){
                resetCounter();
            }
        } catch (SQLException e) {
            throw new DaoException("Failed to initialize table "+ TABLE_NAME,e);
        }
    }

    /**
     * @return the counter's value
     * @throws DaoException if an error occurred while trying to select the value
     */
    @Override
    public Integer selectCounter() throws DaoException {
        String query = String.format("SELECT %s FROM %s;", COLUMN_NAME, TABLE_NAME);
        OfflineResultSet resultSet;
        try {
            resultSet = cursor.executeRead(query);
        } catch (SQLException e) {
            throw new DaoException("Failed to select "+ COLUMN_NAME ,e);
        }
        if(resultSet.next()){
            return resultSet.getInt(COLUMN_NAME);
        } else {
            throw new DaoException("Failed to select "+ COLUMN_NAME);
        }
    }

    /**
     * @param value - new value for the counter
     * @throws DaoException if an error occurred while trying to insert the value
     */
    @Override
    public void insertCounter(Integer value) throws DaoException {
        String query = String.format("DELETE FROM %s; INSERT INTO %s (%s) VALUES (%d);",TABLE_NAME, TABLE_NAME, COLUMN_NAME, value);
        try {
            if(cursor.executeWrite(query) == 0){
                throw new RuntimeException("Unexpected error while inserting " + COLUMN_NAME);
            }
        } catch (SQLException e) {
            throw new DaoException("Failed to insert " + COLUMN_NAME,e);
        }
    }

    /**
     * @throws DaoException if an error occurred while trying to increment the counter
     */
    @Override
    public void incrementCounter() throws DaoException {
        String query = String.format(" UPDATE %s SET %s = (SELECT %s FROM %s) + 1;",
                TABLE_NAME,
                COLUMN_NAME,
                COLUMN_NAME,
                TABLE_NAME);
        try {
            if(cursor.executeWrite(query) != 1){
                throw new RuntimeException("Unexpected error while incrementing " + COLUMN_NAME);
            }
        } catch (SQLException e) {
            throw new DaoException("Failed to increment " + COLUMN_NAME,e);
        }
    }

    /**
     * @throws DaoException if an error occurred while trying to reset the counter
     */
    @Override
    public void resetCounter() throws DaoException {
        insertCounter(1);
    }
}

File: common/src/main/java/com/imsproject/common/dataAccess/abstracts/CounterDAOBase.kt
package com.imsproject.common.dataAccess.abstracts

import com.imsproject.common.dataAccess.*
import java.sql.SQLException


abstract class CounterDAOBase protected constructor(
    private val cursor: SQLExecutor,
    private val tableName: String,
    private val columnName: String
) : CounterDAO {

    /**
     * Initializes the table in the database if it does not exist.
     * @throws DaoException if an error occurred while trying to initialize the table
     */
    @Throws(DaoException::class)
    protected fun initTable() {
        val query = CreateTableQueryBuilder(tableName)
            .addColumn(
                columnName,
                ColumnType.INTEGER,
                modifiers = ColumnModifiers.NOT_NULL
            ).build()
        try {
            cursor.executeWrite(query)
            val query2 = String.format("SELECT * FROM %s;", tableName)
            if (cursor.executeRead(query2).isEmpty) {
                resetCounter()
            }
        } catch (e: SQLException) {
            throw DaoException("Failed to initialize table $tableName", e)
        }
    }

    /**
     * @return the counter's value
     * @throws DaoException if an error occurred while trying to select the value
     */
    @Throws(DaoException::class)
    override fun selectCounter(): Int {
        val query = String.format("SELECT %s FROM %s;", columnName, tableName)
        val resultSet: OfflineResultSet
        try {
            resultSet = cursor.executeRead(query)
        } catch (e: SQLException) {
            throw DaoException("Failed to select $columnName", e)
        }
        if (resultSet.next()) {
            return resultSet.getTyped(columnName) ?: throw DaoException("Failed to select $columnName")
        } else {
            throw DaoException("Failed to select $columnName")
        }
    }

    /**
     * @param value - new value for the counter
     * @throws DaoException if an error occurred while trying to insert the value
     */
    @Throws(DaoException::class)
    override fun insertCounter(value: Int) {
        val query = String.format(
            "DELETE FROM %s; INSERT INTO %s (%s) VALUES (%d);",
            tableName,
            tableName,
            columnName,
            value
        )
        try {
            if (cursor.executeWrite(query) == 0) {
                throw RuntimeException("Unexpected error while inserting $columnName")
            }
        } catch (e: SQLException) {
            throw DaoException("Failed to insert $columnName", e)
        }
    }

    /**
     * @throws DaoException if an error occurred while trying to increment the counter
     */
    @Throws(DaoException::class)
    override fun incrementCounter() {
        val query = String.format(
            "UPDATE %s SET %s = (SELECT %s FROM %s) + 1;",
            tableName,
            columnName,
            columnName,
            tableName
        )
        try {
            if (cursor.executeWrite(query) != 1) {
                throw RuntimeException("Unexpected error while incrementing $columnName")
            }
        } catch (e: SQLException) {
            throw DaoException("Failed to increment $columnName", e)
        }
    }

    /**
     * @throws DaoException if an error occurred while trying to reset the counter
     */
    @Throws(DaoException::class)
    override fun resetCounter() {
        insertCounter(1)
    }
}
